### ISSUE TITLE
[FeatureRequest] Contribution of single camera on SfM (and visa versa)

### ISSUE THREAD

**Issue by Baasje85** [alicevision/meshroom#512]
It would be good to see the influence of a single camera on the point cloud. A rendering I could think of is using a selection colour on the point cloud. The inverse: what cameras have contributed to a single point would also be interesting.

From GUI perspective: 
 - a single click on a camera, would highlight the part of the point cloud influenced
 - a single click on a point, would highlight the cameras contributing to the location

**Comment by laurelkeys** [comment]
Hi, I'm also really interested in this featured.

While there isn't currently a way to see which points are related to which photos on the GUI, may I get some advice on how to get this information from the logs (e.g. which node's info should I be looking into)?

I've just recently started using Meshroom so I'm still a bit unfamiliar with the code structure, and reading throught the docs I couldn't find an answer to this problem, so searching for it was what actually led me to this issue.

I tried looking into the `.log` files from the `FeatureExtraction` to the `StructureFromMotion` nodes in the default pipeline but couldn't really get a grasp of them for the most part. To better understand what each of them does, is there a good place to start reading into?

Thank you.

**Comment by laurelkeys** [comment]
@natowi  I'm using Meshroom to reconstruct images captured by multiple drones (currently simulated on [AirSim](https://github.com/microsoft/AirSim)), with the aim of optimizing their trajectory for maximum scene coverage, so I'm trying to identify which places in the cloud are less dense to capture more pictures in those areas.

Therefore, I'd like to be able to query which points in the reconstructed point cloud came from a specific image, and also which camera viewpoints have contributed to a single point.

I was looking into the `StructureFromMotion` output (converted to JSON) and, from what I understood, it seems like I can get most of this info from there, since it has a `"viewId"` for each image and inside of `"structure"` I can get the coordinates of each point with `"X"`, and which pictures were used with `"observations"`.

Is this the information that would be used to implement the requested feature on the GUI?

**Comment by fabiencastan** [comment]
@laurelkeys We could setup a confcall if you are interested to contribute to the implementation of such a feature.

**Comment by natowi** [comment]
> a single click on a camera, would highlight the part of the point cloud influenced
   a single click on a point, would highlight the cameras contributing to the location



[UAV Mapper](https://github.com/cvBin/UAV-Mapper) has this feature, but sadly only binaries are provided so far. 

![singlepoint](https://user-images.githubusercontent.com/8138537/84590931-ad79d380-ae3a-11ea-9cb1-b5175cb25dc8.JPG)
![coverage](https://user-images.githubusercontent.com/8138537/84590936-b5397800-ae3a-11ea-9f49-32707925eed1.JPG)

**Comment by EmilienCerem** [comment]
@laurelkeys Hello, 

I want to have a precision on what you said. I want in a file, not visually in the gui, the coordinates of the matches points of each pair of photos in the point clouds. I think that the pipeline to create the point clouds use this information but i don't find it in the files... I just want an equivalent of the matches point in pixel in the photos and in (x,y,z) system in the cloud points. I looked what you said but in structure from motion there is only point clouds of intermediate_steps and two other points clouds, the rest of the files are only statistics files, i don't have a file with coordinates of points for each "viewid", where do you find this ?

**Comment by fabiencastan** [comment]
@EmilienCerem You can retrieve it from "observations" on the 3d landmarks in the sfmData.
See:
https://github.com/alicevision/AliceVision/blob/71273a615935d22d3e3beae89124228ca5f944cc/src/aliceVision/sfmData/SfMData.hpp#L64
https://github.com/alicevision/AliceVision/blob/71273a615935d22d3e3beae89124228ca5f944cc/src/aliceVision/sfmData/Landmark.hpp#L64

**Comment by EmilienCerem** [comment]
> @EmilienCerem You can retrieve it from "observations" on the 3d landmarks in the sfmData.
> See:
> https://github.com/alicevision/AliceVision/blob/71273a615935d22d3e3beae89124228ca5f944cc/src/aliceVision/sfmData/SfMData.hpp#L64
> https://github.com/alicevision/AliceVision/blob/71273a615935d22d3e3beae89124228ca5f944cc/src/aliceVision/sfmData/Landmark.hpp#L64

The sfmData is the "sfm.abc" file ?

**Comment by EmilienCerem** [comment]
> @EmilienCerem You can retrieve it from "observations" on the 3d landmarks in the sfmData.
> See:
> https://github.com/alicevision/AliceVision/blob/71273a615935d22d3e3beae89124228ca5f944cc/src/aliceVision/sfmData/SfMData.hpp#L64
> https://github.com/alicevision/AliceVision/blob/71273a615935d22d3e3beae89124228ca5f944cc/src/aliceVision/sfmData/Landmark.hpp#L64

So i have to convert "smf.abc" in json and look the file to see the structure ?

**Comment by fabiencastan** [comment]
Yes, or use the AliceVision API to load the ABC and access the SfMData C++ structures.

**Comment by natowi** [comment]
https://github.com/alicevision/Meshroom/pull/2177
